content/blog/package-react-components/index.md

```diff
@@ -0,0 +1,776 @@
+title: Package, Publish and Distribute React Components in NPM
+date: "2020-01-18"
+author: franleplant
+description:
+  "In this post we explore in detail how to share React Components which are typically made of a
+  bundle of Javascript, CSS (styles), Images, etc.
+  We explore how to package and bundle the code with Webpack so that it can be consumed appropriately,
+  and we explore how to publish and distribute it to NPM so that it can be easily accessible either
+  via the public NPM repository or a private repository.
+  We also analyse different theming strategies for larger Component Libraries and UI Frameworks.
+  And finally we provide a recommended way of packaging your Components in a simple and very flexible way."
+tags:
+  - React
+  - Javascript
+  - Typescript
+  - Css
+  - Webpack
+  - NPM
+  - code sharing
+  - publishing
+---
+
+## Table of Contents
+
+```toc
+# This code block gets replaced with the TOC
+to-heading: 3
+```
+
+## Introduction
+
+> Check [TLDR](#tldr-how-to-package-and-publish-front-end-libraries-in-2020) if you don't want all
+> details about how and why we reach to the conclusion that that is a valid and desirable solution for the
+> problem at hand
+
+There are diverse situation in which we want to create an NPM package that contains
+a [React][1] Component with css and images.
+
+This situation is particularly common inside companies that want to share components,
+and pieces of code in general, between apps usually inside a private NPM repository, but
+it also apply to Open Source software like a single Component package such as a Date Picker
+or a Multi Select and also to Component Libraries or UI Frameworks like [React Bootstrap][9],
+[Ant Design][8], [Semantic UI][4], [Material UI][7], etc (more on these later).
+
+This problem might seem simple enough but it turns out that it has a lot of nuances and trade-offs.
+
+The main challenge in packaging a [React][1] Component is surprisingly not Javascript,
+but all the Assets that might be related to that component such as CSS, Images, SVGs and fonts.
+These assets are directly related to our Javascript code via some of the following statements:
+
+- `imports image from './image.{png,jpg,svg}'` in Javascript
+- `import classNames from './styles.{css,scss,less)'` in Javascript
+- `url(image.{png,jpg,svg})` in CSS
+- `font-family and @import` in CSS
+
+[Webpack][2] can understand all these dependency statements via _loaders_ and completely
+analyse the cross-language and cross-file-type dependency graph, which for a very simple
+component might look something like this.
+
+```
+            --> styles.css --> otherImage.svg
+MyComponent --> image.svg
+
+```
+
+It turns out that there are a lot of schemes for representing this graph in a Packaged NPM module,
+each with it's own trade-offs. Analysing each alternative is the main purpose of this post.
+
+Unfortunately in early 2020 there still no universal standard way of Packaging a Front End module with all
```

Comment:
minor: This is missing the verb ("there _is_ still no..." or whatever)

Reply:
Also when reading it's weird to see the capital P in Packaging because it came up before and it did not have a capital P IIRC.